File: backend/internal/models/parental.go
package models

// ParentalStatus represents the parental control state visible to users.
type ParentalStatus struct {
	PinSet              bool   `json:"pin_set"`
	AdultContentEnabled bool   `json:"adult_content_enabled"`
}

// AdminParentalStatus extends ParentalStatus with admin-visible fields.
type AdminParentalStatus struct {
	PinSet               bool     `json:"pin_set"`
	RestrictedGenreCodes []string `json:"restricted_genre_codes"`
}

// SetPinInput is the request body for setting a parental PIN.
type SetPinInput struct {
	Pin string `json:"pin" binding:"required,min=4,max=6,numeric"`
}

// VerifyPinInput is the request body for verifying a parental PIN.
type VerifyPinInput struct {
	Pin string `json:"pin" binding:"required"`
}

// UpdateRestrictedGenresInput is the request body for updating restricted genre codes.
type UpdateRestrictedGenresInput struct {
	Codes []string `json:"codes" binding:"required"`
}

// SetUserAdultContentInput is the request body for admin to set user adult content access.
type SetUserAdultContentInput struct {
	AdultContentEnabled bool `json:"adult_content_enabled"`
}

// UserAdultStatus represents a user's adult content access status (for admin listing).
type UserAdultStatus struct {
	UserID              string `json:"user_id"`
	Username            string `json:"username"`
	DisplayName         string `json:"display_name"`
	Role                string `json:"role"`
	AdultContentEnabled bool   `json:"adult_content_enabled"`
}

File: frontend/src/api/parental.ts
import api from './client'

export interface ParentalStatus {
  adult_content_enabled: boolean
  pin_set: boolean
}

export interface AdminParentalStatus {
  pin_set: boolean
  restricted_genre_codes: string[]
}

export interface UserAdultStatus {
  user_id: string
  username: string
  display_name: string
  role: string
  adult_content_enabled: boolean
}

// --- User endpoints ---

export async function getMyParentalStatus(): Promise<ParentalStatus> {
  const { data } = await api.get<ParentalStatus>('/me/parental/status')
  return data
}

export async function unlockAdultContent(pin: string): Promise<void> {
  await api.post('/me/parental/unlock', { pin })
}

export async function lockAdultContent(): Promise<void> {
  await api.post('/me/parental/lock')
}

// --- Admin endpoints ---

export async function getAdminParentalStatus(): Promise<AdminParentalStatus> {
  const { data } = await api.get<AdminParentalStatus>('/admin/parental/status')
  return data
}

export async function getRestrictedGenres(): Promise<{ codes: string[] }> {
  const { data } = await api.get<{ codes: string[] }>('/admin/parental/genres')
  return data
}

export async function updateRestrictedGenres(codes: string[]): Promise<void> {
  await api.put('/admin/parental/genres', { codes })
}

export async function setParentalPin(pin: string): Promise<void> {
  await api.post('/admin/parental/pin', { pin })
}

export async function removeParentalPin(): Promise<void> {
  await api.delete('/admin/parental/pin')
}

export async function listUsersAdultStatus(): Promise<UserAdultStatus[]> {
  const { data } = await api.get<UserAdultStatus[]>('/admin/parental/users')
  return data
}

export async function setUserAdultContent(userId: string, enabled: boolean): Promise<void> {
  await api.put(`/admin/parental/users/${userId}`, { adult_content_enabled: enabled })
}
